File: backend/app/services/resume_locale.py
from __future__ import annotations

from copy import deepcopy
import re
from typing import Any


SUPPORTED_RESUME_LANGUAGES = {"zh-CN", "en"}

RESUME_LOCALES: dict[str, dict[str, Any]] = {
    "zh-CN": {
        "html_lang": "zh-CN",
        "unnamed": "未命名",
        "avatar_alt": "头像",
        "colon": "：",
        "list_separator": "、",
        "skill_keywords": "技能标签",
        "project_tech_stack": "技术栈",
        "custom_tech_stack": "技术/工具",
        "experience": "经历",
        "content": "内容",
        "candidate": "求职者",
        "anonymous_resume": "匿名简历",
        "section_titles": {
            "basics": "基本信息",
            "summary": "个人简介",
            "education": "教育经历",
            "skills": "专业技能",
            "work": "工作经历",
            "projects": "项目经历",
            "awards": "荣誉奖项",
        },
        "field_config": {
            "phone": "电话",
            "email": "邮箱",
            "status": "当前状态",
            "location": "地点",
            "highest_degree": "最高学历",
            "website": "个人网站",
            "github": "代码仓库",
            "expected_salary": "期望薪资",
        },
    },
    "en": {
        "html_lang": "en",
        "unnamed": "Untitled",
        "avatar_alt": "Profile photo",
        "colon": ": ",
        "list_separator": " / ",
        "skill_keywords": "Skills",
        "project_tech_stack": "Tech Stack",
        "custom_tech_stack": "Tools",
        "experience": "Experience",
        "content": "Content",
        "candidate": "Candidate",
        "anonymous_resume": "Anonymous Resume",
        "section_titles": {
            "basics": "Contact",
            "summary": "Professional Summary",
            "education": "Education",
            "skills": "Skills",
            "work": "Work Experience",
            "projects": "Project Experience",
            "awards": "Awards",
        },
        "field_config": {
            "phone": "Phone",
            "email": "Email",
            "status": "Status",
            "location": "Location",
            "highest_degree": "Education",
            "website": "Website",
            "github": "GitHub",
            "expected_salary": "Expected Salary",
        },
    },
}


def normalize_resume_language(value: Any) -> str:
    language = str(value or "").strip()
    if language.lower().startswith("en"):
        return "en"
    return "zh-CN"


def detect_resume_language(text: str) -> str:
    value = str(text or "")
    han_count = len(re.findall(r"[\u3400-\u9fff]", value))
    latin_count = len(re.findall(r"[A-Za-z]", value))
    if latin_count >= 40 and latin_count > han_count * 2:
        return "en"
    return "zh-CN"


def _resume_display_text(value: Any) -> list[str]:
    if isinstance(value, str):
        return [value]
    if isinstance(value, list):
        result: list[str] = []
        for item in value:
            result.extend(_resume_display_text(item))
        return result
    if isinstance(value, dict):
        result = []
        ignored_keys = {
            "id",
            "icon",
            "preset_type",
            "avatar",
            "url",
            "website",
            "github",
            "field_config",
            "layout",
        }
        for key, item in value.items():
            if key not in ignored_keys:
                result.extend(_resume_display_text(item))
        return result
    return []


def resolve_resume_language(value: Any, resume_data: Any = None) -> str:
    stored = normalize_resume_language(value)
    if not isinstance(resume_data, dict):
        return stored
    layout = resume_data.get("layout")
    if isinstance(layout, dict) and layout.get("language_locked") is True:
        return stored
    if stored == "en":
        return stored

    titles = layout.get("section_titles") if isinstance(layout, dict) else {}
    title_values = set(titles.values()) if isinstance(titles, dict) else set()
    english_titles = set(RESUME_LOCALES["en"]["section_titles"].values())
    chinese_titles = set(RESUME_LOCALES["zh-CN"]["section_titles"].values())
    english_hits = len(title_values & english_titles)
    chinese_hits = len(title_values & chinese_titles)
    if english_hits >= 2 and english_hits > chinese_hits:
        return "en"

    display_text = "\n".join(_resume_display_text(resume_data))
    return detect_resume_language(display_text)


def resume_locale(language: Any) -> dict[str, Any]:
    return deepcopy(RESUME_LOCALES[normalize_resume_language(language)])


def localized_section_title(key: str, title: Any, language: Any) -> str:
    locale = resume_locale(language)
    localized = locale["section_titles"].get(key)
    current = str(title or "").strip()
    if not localized:
        return current or key
    known_defaults = {
        values["section_titles"].get(key)
        for values in RESUME_LOCALES.values()
        if values["section_titles"].get(key)
    }
    if not current or current == key or current in known_defaults:
        return localized
    return current


def section_field_labels(layout: dict[str, Any], section_key: str, language: Any, *, custom: bool = False) -> dict[str, str]:
    locale = resume_locale(language)
    defaults: dict[str, str] = {}
    if section_key == "projects":
        defaults["tech_stack"] = locale["project_tech_stack"]
    elif section_key == "skills":
        defaults["keywords"] = locale["skill_keywords"]
    elif custom:
        defaults["tech_stack"] = locale["custom_tech_stack"]

    configured = layout.get("field_labels") if isinstance(layout, dict) else {}
    section_config = configured.get(section_key) if isinstance(configured, dict) else {}
    if isinstance(section_config, dict):
        for key, value in section_config.items():
            if isinstance(value, str):
                defaults[key] = value
    return defaults

File: frontend/src/utils/resumeLocale.ts
import type { ResumeData } from "@/types/resume"

export type ResumeLanguage = "zh-CN" | "en"

export const resumeLocales = {
  "zh-CN": {
    sectionTitles: {
      basics: "基本信息",
      summary: "个人简介",
      education: "教育经历",
      skills: "专业技能",
      work: "工作经历",
      projects: "项目经历",
      awards: "荣誉奖项",
    },
    fieldConfig: {
      phone: "电话",
      email: "邮箱",
      status: "当前状态",
      location: "地点",
      highest_degree: "最高学历",
      website: "个人网站",
      github: "代码仓库",
      expected_salary: "期望薪资",
    },
    fieldLabels: {
      projectTechStack: "技术栈",
      customTechStack: "技术/工具",
    },
  },
  en: {
    sectionTitles: {
      basics: "Contact",
      summary: "Professional Summary",
      education: "Education",
      skills: "Skills",
      work: "Work Experience",
      projects: "Project Experience",
      awards: "Awards",
    },
    fieldConfig: {
      phone: "Phone",
      email: "Email",
      status: "Status",
      location: "Location",
      highest_degree: "Education",
      website: "Website",
      github: "GitHub",
      expected_salary: "Expected Salary",
    },
    fieldLabels: {
      projectTechStack: "Tech Stack",
      customTechStack: "Tools",
    },
  },
} as const

export function normalizeResumeLanguage(value: unknown): ResumeLanguage {
  return String(value || "").toLowerCase().startsWith("en") ? "en" : "zh-CN"
}

function displayText(value: unknown): string[] {
  if (typeof value === "string") return [value]
  if (Array.isArray(value)) return value.flatMap(displayText)
  if (!value || typeof value !== "object") return []
  const ignoredKeys = new Set(["id", "icon", "preset_type", "avatar", "url", "website", "github", "field_config", "layout"])
  return Object.entries(value)
    .filter(([key]) => !ignoredKeys.has(key))
    .flatMap(([, item]) => displayText(item))
}

export function resolveResumeLanguage(value: unknown, data?: ResumeData): ResumeLanguage {
  const stored = normalizeResumeLanguage(value)
  if (!data || data.layout?.language_locked === true || stored === "en") return stored
  const titleValues = new Set(Object.values(data.layout?.section_titles || {}))
  const englishTitles = new Set<string>(Object.values(resumeLocales.en.sectionTitles))
  const chineseTitles = new Set<string>(Object.values(resumeLocales["zh-CN"].sectionTitles))
  const englishHits = [...titleValues].filter((title) => englishTitles.has(title)).length
  const chineseHits = [...titleValues].filter((title) => chineseTitles.has(title)).length
  if (englishHits >= 2 && englishHits > chineseHits) return "en"

  const text = displayText(data).join("\n")
  const hanCount = (text.match(/[\u3400-\u9fff]/g) || []).length
  const latinCount = (text.match(/[A-Za-z]/g) || []).length
  return latinCount >= 40 && latinCount > hanCount * 2 ? "en" : "zh-CN"
}

export function defaultSectionTitle(key: string, language: unknown) {
  return resumeLocales[normalizeResumeLanguage(language)].sectionTitles[key as keyof typeof resumeLocales["zh-CN"]["sectionTitles"]]
}

export function defaultFieldConfigLabel(key: string, language: unknown) {
  return resumeLocales[normalizeResumeLanguage(language)].fieldConfig[key as keyof typeof resumeLocales["zh-CN"]["fieldConfig"]]
}

export function effectiveFieldLabel(data: ResumeData, sectionKey: string, fieldKey: string, language: unknown, custom = false) {
  const configured = data.layout.field_labels?.[sectionKey]?.[fieldKey]
  if (typeof configured === "string") return configured
  if (fieldKey !== "tech_stack") return ""
  const locale = resumeLocales[normalizeResumeLanguage(language)]
  return custom ? locale.fieldLabels.customTechStack : locale.fieldLabels.projectTechStack
}

export function applyResumeLanguage(data: ResumeData, previousLanguage: unknown, nextLanguage: unknown) {
  const previous = resumeLocales[normalizeResumeLanguage(previousLanguage)]
  const next = resumeLocales[normalizeResumeLanguage(nextLanguage)]
  const previousTitles = previous.sectionTitles as Record<string, string>
  const nextTitles = next.sectionTitles as Record<string, string>
  data.layout.section_titles ||= {}
  for (const key of Object.keys(next.sectionTitles)) {
    const current = data.layout.section_titles[key]
    const knownDefaults = Object.values(resumeLocales).map((locale) => (locale.sectionTitles as Record<string, string>)[key])
    if (!current || current === key || current === previousTitles[key] || knownDefaults.includes(current)) {
      data.layout.section_titles[key] = nextTitles[key]
    }
  }

  const fieldConfig = data.basics?.field_config
  if (fieldConfig && typeof fieldConfig === "object" && !Array.isArray(fieldConfig)) {
    const nextFieldConfig = next.fieldConfig as Record<string, string>
    for (const key of Object.keys(next.fieldConfig)) {
      const config = fieldConfig[key]
      if (!config || typeof config !== "object" || Array.isArray(config)) continue
      const knownDefaults = Object.values(resumeLocales).map((locale) => (locale.fieldConfig as Record<string, string>)[key])
      if (!config.label || knownDefaults.includes(config.label)) {
        config.label = nextFieldConfig[key]
      }
    }
  }
}
